### PR TITLE
Remove label and placeholder from context and form, add optional provideDefaultLabelFromFieldName property to bridges.

### PR DESCRIPTION
**Pull request by Monteth**
Changes are described here: https://github.com/vazco/uniforms/issues/973#issuecomment-1235430047.

This version doesn't include the `provideDefaultLabelFromFieldName` bridge property in SimpleSchema, since it always delivers a 'label' property. We cannot distinguish whether it was inferred from the field name or entered manually into the schema.

### CODE REVIEW

File: packages/uniforms-bridge-graphql/__tests__/GraphQLBridge.ts [hunk 1]
```diff
@@ -225,12 +224,16 @@ describe('GraphQLBridge', () => {
 
   describe('#getProps', () => {
     describe('labels are derived properly', () => {
-      describe('when props.label is undefined or true', () => {
+      describe('when provideDefaultLabelFromFieldName is true', () => {
+        const bridgeTDefaultLabel = new GraphQLBridge(
+          astT.getType('Post')!,
+          schemaValidator,
+          schemaData,
+        );
```
Comment (radekmie) [inline packages/uniforms-bridge-graphql/__tests__/GraphQLBridge.ts:232]: It's missing the `provideDefaultLabelFromFieldName`, right?

File: packages/uniforms-bridge-graphql/src/GraphQLBridge.ts [hunk 2]
```diff
@@ -125,7 +125,12 @@ export default class GraphQLBridge extends Bridge {
       props.decimal = true;
     }
 
-    props.label ??= upperFirst(lowerCase(field.name));
+    if (
+      this.extras.provideDefaultLabelFromFieldName &&
+      props.label === undefined
+    ) {
+      props.label = upperFirst(lowerCase(field.name));
+    }
```
Comment (radekmie) [inline packages/uniforms-bridge-graphql/src/GraphQLBridge.ts:133]: I think we should make this bridge accept an object now. The thing is, `extras` is a map from field name to its extra props. In that case, if someone will have a `provideDefaultLabelFromFieldName` field (for some reason), it's going to break.

I propose to accept an object _plus_ this new field instead.

File: packages/uniforms-bridge-json-schema/src/JSONSchemaBridge.ts [hunk 3]
```diff
@@ -107,6 +107,7 @@ export default class JSONSchemaBridge extends Bridge {
     public validator: (
       model: UnknownObject,
     ) => ValidatorResult | null | undefined,
+    private provideDefaultLabelFromFieldName: boolean = true,
```
Comment (radekmie) [inline packages/uniforms-bridge-json-schema/src/JSONSchemaBridge.ts:110]: Same as with the `GraphQLBridge` - it should be an object already.

File: packages/uniforms-bridge-zod/src/ZodBridge.ts [hunk 4]
```diff
@@ -29,7 +29,10 @@ function isNativeEnumValue(value: unknown) {
 }
 
 export default class ZodBridge<T extends ZodRawShape> extends Bridge {
-  constructor(public schema: ZodObject<T>) {
+  constructor(
+    public schema: ZodObject<T>,
+    private provideDefaultLabelFromFieldName: boolean = false,
```
Comment (radekmie) [inline packages/uniforms-bridge-zod/src/ZodBridge.ts:34]: Same here. But maybe it'd make sense to have the schema as the first argument and the options object as the second? I'm not sure.

File: packages/uniforms/src/BaseForm.tsx [hunk 5]
```diff
@@ -21,12 +21,11 @@ export type BaseFormProps<Model extends UnknownObject> = {
...
   error: unknown;
   id?: string;
-  label: boolean;
   model: Model;
   modelTransform?: (mode: ModelTransformMode, model: Model) => Model;
   noValidate: boolean;
-  onChange?: (key: string, value: unknown) => void;
-  onSubmit: (model: Model) => void | Promise<unknown>;
+  onChange?: (key: string, value: any) => void;
+  onSubmit: (model: Model) => void | Promise<any>;
```
Comment (radekmie) [inline packages/uniforms/src/BaseForm.tsx:28]: Is this from an incorrect conflict resolution or actual change?

File: packages/uniforms/src/useField.tsx [hunk 6]
```diff
@@ -59,25 +36,12 @@ export function useField<
...
-    state.placeholder,
-    label || labelFallback,
-  );
+  const label = props.label ?? schemaProps.label ?? '';
 
   // eslint-disable-next-line react-hooks/exhaustive-deps
   const id = useMemo(() => context.randomId(), []);
   const onChange = useCallback(
-    (value?: Value, key: string = name) => {
+    (value?: Value | unknown, key: string = name) => {
```
Comment (radekmie) [inline packages/uniforms/src/useField.tsx:44]: Same as the `any` above.

File: website/pages-parts/CommonForms/ShippingForm.tsx [hunk 7]
```diff
@@ -21,7 +21,6 @@ export function ShippingForm() {
       schema={schema}
       onSubmit={(model: any) => alert(JSON.stringify(model, null, 2))}
       model={model}
-      placeholder
```
Comment (radekmie) [inline website/pages-parts/CommonForms/ShippingForm.tsx:24]: We'd like to add them in the schema then.

**Review by radekmie — CHANGES_REQUESTED** [review]
As I understand, as soon as we migrate to Zod in tests _or_ Simple Schema solves the issue with the `label`s, we could remove all of these `label: ''` from tests?

File: packages/uniforms-bridge-graphql/src/GraphQLBridge.ts (hunk 2 above)
Comment (Monteth) [reply packages/uniforms-bridge-graphql/src/GraphQLBridge.ts:133]: Do you want it to look like this?
```ts
constructor(
  public schema: GraphQLType,
  props: {
    validator: (model: UnknownObject) => unknown,
    provideDefaultLabelFromFieldName: boolean
  }
)
```
Or do you want the `schema` to be in the object too?

Comment (Monteth) [reply packages/uniforms-bridge-graphql/src/GraphQLBridge.ts:133]: OK, I can see from the next comments that you are not sure; I would consider dividing parameters to 1. `schema` and 2. `others` if the second parameter would be unused in most cases. In this case of Graphql schema, I think the validator will be used in most cases, the same for JSONSchema, so let's change those bridges to accept one object.

**Comment by Monteth** [comment]
> As I understand, as soon as we migrate to Zod in tests or Simple Schema solves the issue with the labels, we could remove all of these label: '' from tests?

Correct. Since we get rid of `label` in context, we can no longer tell SimpleSchema not to derive the label from the field name, so we have to pass it explicitly. This behavior is specific to SimpleSchema only.

File: website/pages-parts/CommonForms/ShippingForm.tsx (hunk 7 above)
Comment (Monteth) [reply website/pages-parts/CommonForms/ShippingForm.tsx:24]: I don't know what I am doing wrong, but the form behaves exactly like on [uniforms.tools](https://uniforms.tools/). The only empty fields `state` and `zip` receives the placeholder in the field level, so it seems like we don't need it.

File: packages/uniforms-bridge-json-schema/src/JSONSchemaBridge.ts [hunk 8]
```diff
@@ -287,8 +297,10 @@ export default class JSONSchemaBridge extends Bridge {
       this._compiledSchema[name],
     );
 
-    props.label ??=
-      props.title ?? upperFirst(lowerCase(joinName(null, name).slice(-1)[0]));
+    if (this.provideDefaultLabelFromFieldName && props.label === undefined) {
+      props.label =
+        props.title ?? upperFirst(lowerCase(joinName(null, name).slice(-1)[0]));
+    }
```
Comment (radekmie) [inline packages/uniforms-bridge-json-schema/src/JSONSchemaBridge.ts:303]:
```suggestion
    props.label ??= props.title;
    if (this.provideDefaultLabelFromFieldName && props.label === undefined) {
      props.label = upperFirst(lowerCase(joinName(null, name).slice(-1)[0]));
    }
```

File: packages/uniforms-bridge-graphql/src/GraphQLBridge.ts [hunk 9]
```diff
@@ -20,13 +20,29 @@ function fieldInvariant(name: string, condition: boolean): asserts condition {
...
-    public extras: UnknownObject = {},
-  ) {
+  extras: UnknownObject;
+  provideDefaultLabelFromFieldName: boolean;
+  schema: GraphQLType;
+  validator: (model: UnknownObject) => unknown;
+
+  constructor({
+    extras = {},
+    provideDefaultLabelFromFieldName = false,
```
Comment (radekmie) [inline packages/uniforms-bridge-graphql/src/GraphQLBridge.ts:30]:
```suggestion
    provideDefaultLabelFromFieldName = true,
```

**Review by radekmie — DISMISSED** [review]

File: packages/uniforms/__tests__/connectField.tsx [hunk 10]
```diff
@@ -206,25 +190,28 @@ describe('connectField', () => {
...
-    expect(wrapper.find(Test).at(4).prop('label')).toBe('');
-    expect(wrapper.find(Test).at(5).prop('label')).toBe('');
+    expect(wrapper.find(Test).at(0).prop('label')).toBe('Field');
+    expect(wrapper.find(Test).at(1).prop('label')).toBe('');
+    expect(wrapper.find(Test).at(2).prop('label')).toBe('Field');
+    expect(wrapper.find(Test).at(3).prop('label')).toBe('Other');
+    expect(wrapper.find(Test).at(4).prop('label')).toBe('Subfield');
+    expect(wrapper.find(Test).at(5).prop('label')).toBe('Subfield');
     expect(wrapper.find(Test).at(6).prop('label')).toBe('Subfield');
+    expect(wrapper.find(Test).at(7).prop('label')).toBe('Subfield');
```
Comment (kestarumper) [inline packages/uniforms/__tests__/connectField.tsx:215]: I counted 9 fields but I see only 8 `expect`(s). I was looking for
```ts
expect(wrapper.find(Test).at(8).prop('label')).toBe('');
```

**Review by kestarumper — DISMISSED** [review]

**Review by kestarumper — APPROVED** [review]

**Review by radekmie — APPROVED** [review]

**Pull request merged**